modules/contentbox/models/security/RateLimiter.cfc

```diff
@@ -42,99 +35,77 @@ component extends="coldbox.system.Interceptor" {
 		}
 	}
 
-	/**
-	 * Written by Charlie Arehart, charlie@carehart.org, in 2009, updated 2012, modified by Luis Majano 2016
-	 * - Throttles requests made more than "count" times within "duration" seconds from single IP.
-	 * - Duck typed for performance
-	 *
-	 * @count    The throttle counter
-	 * @duration The time in seconds to limit
-	 * @event    The request context object
-	 * @settings The settings structure
-	 */
-	private function limiter( count, duration, event, settings ){
+	/** 
+	* Written by Charlie Arehart, charlie@carehart.org, in 2009, updated 2012, modified by Luis Majano 2016
+	* - Throttles requests made more than "count" times within "duration" seconds from single IP. 
+	* - Duck typed for performance
+	* @count	The throttle counter
+	* @duration	The time in seconds to limit
+	* @event	The request context object
+	* @settings	The settings structure
+	*/
+	private function limiter( count, duration, event, settings ) {
 		// Get real IP address of requester
-		var realIP = variables.securityService.getRealIP();
+		var realIP 		= variables.securityService.getRealIP();
+		var cache 		= cachebox.getDefaultCache();
+		var cacheKey 	= 'limiter'&realIP;
 
 		// If first time visit, create record.
-		if ( !structKeyExists( variables.limitData, realIP ) ) {
-			lock name="cb-ratelimiter-#hash( realIP )#" type="exclusive" throwontimeout="true" timeout="5" {
-				if ( !structKeyExists( variables.limitData, realIP ) ) {
-					variables.limitData[ realIP ] = { attempts : 1, lastAttempt : now() };
-					return this;
-				}
-			}
-		}
-
-		lock name="cb-ratelimiter-#hash( realIP )#" type="readonly" throwontimeout="true" timeout="5" {
-			var targetData = variables.limitData[ realIP ];
+		var targetData = cache.get( cacheKey );
```

Comment:
best to use the `getOrSet()` for these type of situations.https://s3.amazonaws.com/apidocs.ortussolutions.com/cachebox/7.4.0/cachebox/system/cache/AbstractCacheBoxProvider.html#getOrSet()

Reply:
Like this? But how to handle, if cache key doesn't exist return default value (no further execution!
```
		var targetData = cache.getOrSet( cacheKey, function(){
				return { attempts = 0, lastAttempt = now() }
		} );
```

Reply:
What this does is pretty much what you did manually.  If the cacheKey exists, it returns the value of the cache. If it doesn't, it will seed it with the return value of the supplier function.

Reply:
yes, got it, but in case value is created I don't want to execute further code. Original code did this: 
`		if( isNull( targetData ) ){
			cache.set( cacheKey, { attempts = 1, lastAttempt = now() });
			return this;
		}`

Reply:
Ahh ok. Well, you would say
```
if( targetData.attempts == 1 ){
  return this;
```

The usage of `getOrSet()` is preferred as it does internal locking

Reply:
I get the locking mechanism, to make it working I have to write something like this. 
`		var targetData = cache.getOrSet( cacheKey, function(){
				return { attempts = 0, lastAttempt = now() }
		} );

		// on first visit no further processing
		if( targetData.attempts == 0 ){
			targetData.attempts++;
			cache.set( cacheKey, targetData );
  			return this;
		}`